Nicknames

danielamormocea opened the pull request:
Draft for #1197


@ianb Hi, I wanted you to review what I've been at so far, as I'm not so familiar with React and it's my first time interacting with it and I don't know if I'm on the right track here; so I kindly ask you to give me some feedback for these changes (this is not how I want it to look, my idea was actually very similar to what @awallin  proposed), I just made something schematic so I could get some feedback based on this. Thank you!

![firefoxv](https://user-images.githubusercontent.com/38378382/76459074-36b8de80-63e4-11ea-8399-c5672eeac87c.png)

ianb commented on extension/options/optionsView.jsx:
```diff
@@ -35,17 +57,77 @@ const LeftSidebar = ({ version }) => {
+            <button
+              className="tab-button styled-button"
+              onClick={() => clickTab(TABS.NICKNAMES)}
+            >
```
What I'd actually prefer here is to just use:

```html
<a className="tab-button" href="#nicknames">Nicknames</a>
```

And then use [onhashchange](https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onhashchange) to catch the change. This way the URL reflects the location, the back button will work, you can bookmark it, etc. Also the hash should be checked on page load.

One thing I'm a little unhappy with in the current code is that `OptionsController` uses state internally, and it's not easy to get state in from the outside. But adding a hashchange handler inside the function isn't good, since the function can (I believe) get called multiple times as state changes.

An example of the solution I came up with is `onKeyboardShortcutError` – it's not a great solution, but at least if you use the same solution then they will be not-great in the same way.

ianb commented on extension/options/optionsView.jsx:
```diff
@@ -3,29 +3,51 @@
         userSettings={userSettings}
         updateUserSettings={updateUserSettings}
-      />
+        active={tabValue === TABS.GENERAL}
```
Instead of this I'd rather this function do:

```js
{tabValue === TABS.GENERAL ?
  <General ... />
  : null}
```

That is, it shouldn't be the component's responsibility to null itself out, instead have this include only the right components.

danielamormocea commented:
Ok. Thanks a lot for all the advice :)

danielamormocea commented:
Hi @ianb @awallin   I've added the rest of the functionalities for this task; When you have a bit of time, please have a look;

I do not display any message for the errors yet, i just check that the inputs are valid (when not, I take no action);


![nickname](https://user-images.githubusercontent.com/38378382/76634448-ee69ff80-654e-11ea-8893-fdeccd2da625.gif)

danielamormocea commented on extension/options/routinesView.jsx:
```diff
@@ -0,0 +1,322 @@
+        <button
+          className="styled-button modal-button"
+          onClick={() => {
+            save();
```
@ianb it doesn't work if the utterance is not a valid intent

danielamormocea replied:
I'm not really sure why it isn't working for you; for me it always works when I introduce a valid intent; the check is being made in the updateNickname function in optionsController.jsx

danielamormocea commented:
@ianb I don't exactly understand where the edit part is going to show. I made the modal because I thought it was easier to interact with; What should I use instead of this modal? A different page?

danielamormocea commented on extension/options/routinesView.jsx:
```diff
@@ -0,0 +1,322 @@
+      <button
+        className="no-style-button"
+        onClick={() => {
+          addUtterance();
```
@ianb are you sure that when introducing a valid intent it doesn't work? I am checking for the validity of the utterance before adding it; (I don't get any errors)

danielamormocea commented on extension/background/main.js:
```diff
@@ -91,6 +92,18 @@ browser.runtime.onMessage.addListener(async (message, sender) => {
+    return intentRunner.getRegisteredNicknames();
+  } else if (message.type === "registerNickname") {
+    let context = message.context;
+    if (context !== null) {
```
@ianb if it's null, the nickname is deleted (from what I've figured from the nickname.js)

ianb commented:
Well, looking at what @awallin put up:

<img width="599" alt="image" src="https://user-images.githubusercontent.com/44390/76664710-3be86b80-6553-11ea-9008-d41f1c37fe30.png">

I don't think you made a menu for edit and remove? I wasn't able to tell exactly. If you haven't already done that, then I would avoid that, it's a very finicky widget and instead it could be two buttons.

I imagine the card expanding with the edit interface in it, and then collapsing when you save.

There won't be any icons for individual nicknames.

danielamormocea commented:
@ianb I followed the model @awallin  proposed, except for the icons (added also the menu and the 2 buttons for edit and remove); I added a gif in a comment above that shows all the flow of the task..I can remove it if is too finicky... and add two button in the right, instead;

danielamormocea commented:
@ianb Also, I noticed a bug; When calling a nickname - if the first intent is "open google" the other ones will not execute; I could make an issue for that if this reproduces

danielamormocea commented:
@ianb I made the changes requested (the errors are now logged in the console); Also, I kept the menu dropdown because I had already implemented it and I think it works well, if you want it changed, I'll see to it;

Please do have a look :)

Summary of all the implemented functionalities:

- Added 2 tabs - General and Routines with additional icons and highlightings on tab change;
- Added list of existing routines with name and description (that being the actions of the routines);
- Added button for creating a New Routine; This expands a div where you can insert a nickname and add the actions for it; If they are not valid, it won't let the user Save the changes and an error is shown in the console;
- Added the 3 dotted menu dropdown, from which when clicking on 
     -> Edit it displays again an expanded div with the nickname and actions already filled, and where you can continue editing the routine;
    -> Remove - it removes the routine;


Details regarding implementetation:

- Used the useToggle method to implement the dropdowns and the collapsable div in order to be able to close them when clicking outside of the respective area, or when clicking 'ESCAPE' (for the collapsable div just the esc). 
This method could be used for toggling from now on;
- Used editNicknameDraft for creating the collapsable div to be able to clear the contents when existing the div; 
  _I performed some copies of objects when necessary to break references in react;_
- Used updateNickname method to register / delete and perform checks on the user input regarding creating a new nickname;

ianb commented:
This is looking great! I have another PR I have to look at with some overlapping functionality before getting this landed.

ianb commented:
Followup:  #1310
